TACC/Frontera-Portal
# task/FP-764 Wrap up onboarding epic

## Overview: ##

## PR Status: ##


## Related Jira tickets: ##

* [FP-669](https://jira.tacc.utexas.edu/browse/FP-669)

## Summary of Changes: ##

- Add Onboarding mechanism
- Add webhook callback mechanism
- Changed settings.py to have onboarding steps: MFA, Request Access, Allocation Refresh, Create Systems (requires push keys modal). These steps are mostly there for testing purposes and can be removed.
- Provisional onboarding UI for users based on UI design
- Add keyservice based system creation mechanism and onboarding step, but this is not enabled due to Frontera and Longhorn not supporting keyservice yet
- Users who are not setup_complete will be redirected to onboarding component when attempting to access the workbench. (This does not yet match UI design's intention of showing a 'disabled' workbench with the onboarding UI as a page)
- Users who are not setup_complete will be shown tickets and account profile components that does not use the workbench

## Testing Steps: ##

1. In the django shell, delete your user
2. Restart all containers, especially celery
3. Log in, you should be directed to onboarding setup page
4. You will be prompted to "request access". Click the action link, it should put it into a "waiting for staff approval state"
5. In the shell, give your user the is_staff = True flag, reload the page
6. You should now see Approve and Deny actions on that step. Approve your access.
7. Everything will automatically process, and eventually you will see the Access Dashboard button

## UI Photos:

![97493410-e5d60600-1932-11eb-89a9-3bb7955b6ffc](https://user-images.githubusercontent.com/17181582/97632205-31a2b100-1a00-11eb-9ad9-9bb65c29815f.png)

## Notes: ##

Required group_vars:
```
_WH_BASE_URL: <should match the domain>
_PORTAL_KEY_SERVICE_ACTOR_ID:  "jzQP0EeX7mE1K"
_PORTAL_USER_ACCOUNT_SETUP_STEPS: [
    {
        'step': 'portal.apps.onboarding.steps.mfa.MFAStep',
        'settings': {}
    },
    {
        'step': 'portal.apps.onboarding.steps.project_membership.ProjectMembershipStep',
        'settings': {
            'project_sql_id': 39726
        }
    },
    {
        'step': 'portal.apps.onboarding.steps.allocation.AllocationStep',
        'settings': {}
    },
    {
        'step': 'portal.apps.onboarding.steps.system_creation.SystemCreationStep',
        'settings': {}
    }
]
```
- settings.py needs to be modified for prod deployment. What should be the default set of onboarding steps? Minimum is allocation and system creation, but request access and project membership should be removed before merge to master. (they're there for testing.)
- Have not yet built an admin interface for listing all users onboarding states, but you can still perform admin actions for individual users (waiting on UI design)
- Admins can perform onboarding admin actions by visiting the user's onboarding page directly at the /onboarding/setup/username route. (Admin is determined by is_staff = True in django)
